### ISSUE TITLE
Warning: "Specified kernel cache directory could not be created"

### ISSUE THREAD

**Issue by adeandrade** [pytorch/pytorch#74415]
### 🐛 Describe the bug

I am getting this warning with the newest version of PyTorch (1.11.0+cu113). In the case shown, we are running the program inside a container that has access to the cache directory mentioned.

I am getting the same warning on different environments that use different cache directories, and all are writeable by the user running the program.
```
/opt/python/lib/python3.9/site-packages/redacted/redacted.py:577: UserWarning: Specified kernel cache directory could not be created!
This disables kernel caching. Specified directory is /root/.
    return half * torch.erfc(const * inputs) 
```

### Versions

1.11.0+cu113

cc @mruberry @ngimel

**Comment by ptrblck** [comment]
I can reproduce the warning in different setups, e.g. the [pytorch docker containers](https://hub.docker.com/r/pytorch/pytorch).
The root cause seems to be in [this line](https://github.com/pytorch/pytorch/blob/c064ea07f6263090a40b1a6fc5fe5973a76785e5/aten/src/ATen/native/cuda/jit_utils.cpp#L856) as `mkdir` is not creating folders recursively and will warn if:
* `$HOME/.cache` is missing (as is the case in the docker image)
* even if `$HOME/.cache` is valid, trying to create `torch/kernels` would still warn:

```
root@3493e1323029:/workspace/src# ls /root/.cache/
pip

root@3493e1323029:/workspace/src# python tmp.py 
tmp.py:13: UserWarning: Specified kernel cache directory could not be created! This disables kernel caching. Specified directory is /root/.cache/torch/kernels. This warning will appear only once per process. (Triggered internally at  ../aten/src/ATen/native/cuda/jit_utils.cpp:860.)
  print(half * torch.erfc(const * inputs))

root@3493e1323029:/workspace/src# mkdir /root/.cache/torch
root@3493e1323029:/workspace/src# python tmp.py 
tensor([[-1.9883]])
tensor([[0.8171]], device='cuda:0')

root@3493e1323029:/workspace/src# ls /root/.cache/torch/
kernels
```

As a workaround you could use `PYTORCH_KERNEL_CACHE_PATH=your_cache_path python script.py` (or `export` this env var) or you could also create the directory directly via `mkdir -p`.

I also think that the Windows version [here](https://github.com/pytorch/pytorch/blob/c064ea07f6263090a40b1a6fc5fe5973a76785e5/aten/src/ATen/native/cuda/jit_utils.cpp#L854) would hit the same issue.
From the docs of [`_mkdir`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/mkdir-wmkdir?view=msvc-170):
> _mkdir can create only one new directory per call, so only the last component of dirname can name a new directory.

**Comment by mruberry** [comment]
Hey @adeandrade -- sorry you're hitting this issue. Is the error message you're getting copied exactly? Are you setting the cache directory to /root, and /root exists, and you're still hitting this error? How are you setting the cache directory, if so?

**Comment by felixdivo** [comment]
I think the problem on Windows is slightly different. It seems like tha path is not assembled correctly (mixing `/` and `\`):

```
UserWarning: Specified kernel cache directory could not be created!
This disables kernel caching.
Specified directory is C:\Users\<MyUserName>\AppData\Local\Temp/torch/kernels.
This warning will appear only once per process.
(Triggered internally at  C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\cuda\jit_utils.cpp:860.)
```

[Added line breaks for readability]

**Comment by mruberry** [comment]
> I think the problem on Windows is slightly different. It seems like tha path is not assembled correctly (mixing `/` and `\`):
> 
> ```
> UserWarning: Specified kernel cache directory could not be created!
> This disables kernel caching.
> Specified directory is C:\Users\<MyUserName>\AppData\Local\Temp/torch/kernels.
> This warning will appear only once per process.
> (Triggered internally at  C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\cuda\jit_utils.cpp:860.)
> ```
> 
> [Added line breaks for readability]

cc @ptrblck

**Comment by syed-ahmed** [comment]
@felixdivo Could you please check if you are still encountering this warning on pytorch nightly binary? I was able to reproduce your case on `1.11.0+cu113` (see below) and am seeing the correct behavior on `1.12.0.dev20220415+cu113`. 

Following are my findings for reference:
- Code behavior is it first looks for PYTORCH_KERNEL_CACHE_PATH, and then TEMP and then HOME on windows.
- set/unset the env vars one by one to see if a warning arises or not.
- Repro script:
  ```python
  import torch
  
  
  half = torch.randn(10, 10, dtype=torch.float16, device='cuda')
  const = torch.ones(10, 10, device='cuda')
  inputs = torch.randn(10, 10, device='cuda')
  
  print(half * torch.erfc(const * inputs))
  ```
- 1.11.0 behavior: https://gist.github.com/syed-ahmed/4a3bfd7bac419cc705d94d1bc55a990c
- nightly behavior: https://gist.github.com/syed-ahmed/6d9cb2281637999428b9ccbb75051ebf

**Comment by felixdivo** [comment]
> @felixdivo Could you please check if you are still encountering this warning on pytorch nightly binary?

Sure, @syed-ahmed.

With `1.11.0+cu113` your script prints the warning on my machine too. I then installed `1.12.0.dev20220416` `py3.9` `cuda11.3` (the most recent one from pip) and the warning does not show any more.

**Comment by syed-ahmed** [comment]
Thanks @felixdivo . Therefore @ptrblck's PR has fixed this issue for windows as well and is on nightly pytorch. You could wait for the next pytorch release or use the nightly pytorch binary or manually set the PYTORCH_KERNEL_CACHE_PATH in 1.11.0 (e.g. `set PYTORCH_KERNEL_CACHE_PATH=C:\Users\terrybel\pytorch_ws_triage\`).

**Comment by felixdivo** [comment]
Thank you very much, now I'll get this to work. 😄

**Comment by shaperilio** [comment]
It seems as of Pytorch 2.10.0, this still happens, i.e. the full cache path must exist, the library itself won't do it.

I am running with `XDG_CACHE_HOME` set to `/cache` (I'm running in a docker container as root so permissions are not an issue). `/cache` definitely exists. The warning says
```
Specified kernel cache directory could not be created! This disables kernel caching. Specified directory is /cache/torch/kernels. This warning will appear only once per process. (Triggered internally at /pytorch/aten/src/ATen/native/cuda/jit_utils.cpp:1487.)
```

And it seems that this `/cache/torch/kernels` exists and has files in it, so *something* is working.
